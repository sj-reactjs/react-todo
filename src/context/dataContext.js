import React, { createContext, useState } from 'react'
import { defaultData } from './../mock/initial-data'

export const TodoDataContext = createContext()

function DataContextProvider(props) {
    const { children } = props;

    const [todoData, setTodoData] = useState(defaultData)
    const [formMode, setformMode] = useState('add')
    const [formData, setFormData] = useState({})

    function addTodo(todo) {
        setTodoData(currentTodo => {
            return [
                ...currentTodo,
                todo
            ]
        })
    }

    function removeTodo(id) {
        setTodoData(currentTodo => currentTodo.filter(todo => todo.id !== id))
    }

    function updateTodo(updatedTodo) {
        setTodoData((currentTodo) => {
            return currentTodo.map(todo => {
                if (todo.id === updatedTodo.id) {
                    return {
                        ...todo,
                        ...updatedTodo,
                    }
                }
                return todo
            })
        })
    }

    function setFormStage(stage = 'add', todoId) {
        switch (stage) {
            case 'edit':
                const itemToEdit = todoData.filter(({ id }) => `${todoId}` === `${id}`)
                if (itemToEdit.length) {
                    setformMode('edit')
                    setFormData(itemToEdit[0])
                }
                break;
            default:
                setformMode('add')
                setFormData({})
                break;
        }
    }

    return (
        <TodoDataContext.Provider value={{
            formMode,
            formData,
            todoData,
            add: addTodo,
            remove: removeTodo,
            update: updateTodo,
            setStage: setFormStage
        }}>
            {children}
        </TodoDataContext.Provider>
    )
}

export { DataContextProvider }
