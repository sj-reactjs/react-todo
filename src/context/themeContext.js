import React, { createContext, useState } from 'react'

const ThemeContext = createContext()

function ThemeContextProvider(props) {
    const { children } = props;
    const defaultTheme = 'dark'
    const [theme, setTheme] = useState(defaultTheme)

    function toggeleTheme() {
        setTheme(currentTheme => currentTheme === 'dark' ? 'light' : 'dark')
    }

    return (
        <ThemeContext.Provider value={{
            theme,
            changeTheme: toggeleTheme
        }}>
            <div className={`todo-app theme-${theme}`}>
                {children}
            </div>
        </ThemeContext.Provider>
    )
}

export { ThemeContextProvider, ThemeContext }
