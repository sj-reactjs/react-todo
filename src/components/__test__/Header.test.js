import { act, render, screen } from '@testing-library/react'
import { Header } from './../index'
import { ThemeContextProvider } from './../../context/themeContext'

describe(`Header component`, () => {
    // not getting button text, need to check 
    it.skip(`render`, () => {
        const { asFragment, container } = render(<ThemeContextProvider><Header /></ThemeContextProvider>)
        const themeButton = container.querySelector('button')
        const buttonText = themeButton.innerHTML
        console.log('aa', themeButton.innerText)
        expect(asFragment()).toMatchSnapshot()
        act(() => {
            themeButton.click()
        })
        const themeButtonUpdatedText = themeButton.innerText
        console.log('bb', themeButton.innerText)
        expect(buttonText).not.toEqual(themeButtonUpdatedText)
    })
})