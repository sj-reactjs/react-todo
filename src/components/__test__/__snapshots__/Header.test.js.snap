// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component render 1`] = `
<DocumentFragment>
  <div
    class="todo-app theme-dark"
  >
    <header
      class="header"
    >
      <button>
        Switch To Light
      </button>
    </header>
  </div>
</DocumentFragment>
`;
